File: LeetCode/Problems/206.py
# NeetCode - Linked List
# 206. Reverse Linked List
# Difficulty : Easy
# Algorithm : Linked List
# Time complexity : O(N), Space complexity : O(N)
# Runtime : 48 ms (80.36%), Memory : 18.52 MB (20.01%)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        res = None
        
        while head:
            linked = ListNode()
            linked.next = res
            linked.val = head.val

            res = linked
            head = head.next

        return res

File: LeetCode/Problems/206.java
/*
206. Reverse Linked List
Difficulty : Easy
Algorithm : Linked List
Runtime : 0 ms (100%), Memory : 42.21 MB (69.84%)
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode cur = head;
        while(cur != null && cur.next != null) {
            ListNode node = cur.next;
            cur.next = cur.next.next;
            node.next = head;
            head = node;
        }
        return head;
    }
}
